### PR TITLE
Fix multipart upload pause

### PR DESCRIPTION
**Pull request by raphkim**
*Issue #, if available:*
#1500 

*Description of changes:*
`TransferUtility#pause(int)` was not behaving correctly when uploading a multipart (> 5MB) item. This was due to interruption detection (due to pause) code returning early before actually canceling the individual `UploadPartTask` tasks in the upload. The same result would have been seen for `TransferUtility#cancel(int)` for a large upload.

Additionally to fixing the above bug, this PR also contains fix to mitigate the effects of race condition observed when pausing multipart upload and resuming it instantly. Previously, the transfer state would be set to `PAUSED` or `CANCELED` as soon as the method was called even before all of the task was interrupted. Now `PENDING_PAUSE` and `PENDING_CANCEL` (already present in the codebase but unused) states are used to make sure that the task is successfully interrupted before the state is updated to its final state.

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: aws-android-sdk-s3-test/src/androidTest/java/com/amazonaws/mobileconnectors/s3/transferutility/PauseTransferIntegrationTest.java [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+import org.junit.Test;
+
+import java.io.File;
+import java.util.Date;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.TimeUnit;
+
+public final class PauseTransferIntegrationTest extends S3IntegrationTestBase {
+
+    private static final String bucketName = "amazon-transfer-util-integ-test-" + new Date().getTime();
```
Comment (desokroshan) [inline aws-android-sdk-s3-test/src/androidTest/java/com/amazonaws/mobileconnectors/s3/transferutility/PauseTransferIntegrationTest.java]: Can we name it like `BUCKET_NAME` to be consistent with convention and rest of the codebase?

File: aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/DownloadTask.java [hunk 2]
```diff
@@ -144,8 +144,6 @@ public Boolean call() {
                      */
                     updater.updateState(download.id, TransferState.WAITING_FOR_NETWORK);
                     LOGGER.debug("Network Connection Interrupted: " + "Moving the TransferState to WAITING_FOR_NETWORK");
-                    ProgressEvent resetEvent = new ProgressEvent(0);
-                    resetEvent.setEventCode(ProgressEvent.RESET_EVENT_CODE);
```
Comment (desokroshan) [inline aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/DownloadTask.java]: Shouldn't we be resetting the progress upon loss of network and then pass `reserEvent` to the `progressListener`?

File: aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/UploadTask.java [hunk 3]
```diff
@@ -310,8 +315,6 @@ private Boolean uploadSinglePartAndWaitForCompletion() {
                      */
                     updater.updateState(upload.id, TransferState.WAITING_FOR_NETWORK);
                     LOGGER.debug("Network Connection Interrupted: " + "Moving the TransferState to WAITING_FOR_NETWORK");
-                    ProgressEvent resetEvent = new ProgressEvent(0);
-                    resetEvent.setEventCode(ProgressEvent.RESET_EVENT_CODE);
```
Comment (desokroshan) [inline aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/UploadTask.java]: Same as above. If `eventCode` is not being used anywhere should we consider removing it altogether from the `ProgressEvent` class?

File: aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/DownloadTask.java [hunk 4]
```diff
@@ -114,16 +114,16 @@ public Boolean call() {
...
+                updater.updateState(download.id, TransferState.CANCELED);
+                LOGGER.info("Transfer is " + TransferState.CANCELED);
                 return false;
             }
 
             // Reset the progress when the transfer is paused.
-            if (TransferState.PAUSED.equals(download.state)) {
-                LOGGER.info("Transfer is " + download.state);
-                ProgressEvent resetEvent = new ProgressEvent(0);
-                resetEvent.setEventCode(ProgressEvent.RESET_EVENT_CODE);
```
Comment (desokroshan) [inline aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/DownloadTask.java]: Are we not resetting on pause on anymore? If so, do we need to update documentation?

Comment (raphkim) [reply aws-android-sdk-s3/src/main/java/com/amazonaws/mobileconnectors/s3/transferutility/DownloadTask.java]: if you look closely, the lines i removed were noops. Im not sure if the original intent was to pass in a `new ProgressEvent(0)` with reset flag, but this object never gets used and a new progress event without flag is reinstantiated instead. Perhaps a different bug, but I didn't want to introduce yet another behavior change in this PR.

File: aws-android-sdk-s3-test/src/androidTest/java/com/amazonaws/mobileconnectors/s3/transferutility/PauseTransferIntegrationTest.java (hunk 1 above)
Comment (raphkim) [reply aws-android-sdk-s3-test/src/androidTest/java/com/amazonaws/mobileconnectors/s3/transferutility/PauseTransferIntegrationTest.java]: yea oops. A mistake that was copy + pasted from my previous mistake. will fix them

**Comment by raphkim** [comment]
This PR contained removal of some lines that SEEMED to be doing something important but was actually not doing anything:

```
ProgressEvent resetEvent = new ProgressEvent(0);
resetEvent.setEventCode(ProgressEvent.RESET_EVENT_CODE);
progressListener.progressChanged(new ProgressEvent(0));
```

The first two lines don't do anything, since `resetEvent` is not used anywhere after this bit of code. 

This is probably an important issue of its own that requires its own investigation, so I decided to leave the redundant lines for now to revisit in the future with a separate PR.

**Pull request merged**